File: zig/src/day05.zig
const std = @import("std");

const Range = struct {
    start: u64,
    end: u64,

    fn contains(self: Range, id: u64) bool {
        return id >= self.start and id <= self.end;
    }
};

fn parseInput(allocator: std.mem.Allocator, input: []const u8) !struct { ranges: []Range, ids: []u64 } {
    var parts = std.mem.splitSequence(u8, std.mem.trim(u8, input, " \t\r\n"), "\n\n");

    const ranges_part = parts.next() orelse return error.InvalidInput;
    const ids_part = parts.next() orelse return error.InvalidInput;

    // Parse ranges
    var ranges = try std.ArrayList(Range).initCapacity(allocator, 10);
    errdefer ranges.deinit(allocator);

    var range_lines = std.mem.splitScalar(u8, ranges_part, '\n');
    while (range_lines.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r");
        if (trimmed.len == 0) continue;

        var parts_iter = std.mem.splitScalar(u8, trimmed, '-');
        const start_str = parts_iter.next() orelse return error.InvalidRange;
        const end_str = parts_iter.next() orelse return error.InvalidRange;

        const start = try std.fmt.parseUnsigned(u64, start_str, 10);
        const end = try std.fmt.parseUnsigned(u64, end_str, 10);

        try ranges.append(allocator, .{ .start = start, .end = end });
    }

    // Parse IDs
    var ids = try std.ArrayList(u64).initCapacity(allocator, 10);
    errdefer ids.deinit(allocator);

    var id_lines = std.mem.splitScalar(u8, ids_part, '\n');
    while (id_lines.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r");
        if (trimmed.len == 0) continue;

        const id = try std.fmt.parseUnsigned(u64, trimmed, 10);
        try ids.append(allocator, id);
    }

    return .{
        .ranges = try ranges.toOwnedSlice(allocator),
        .ids = try ids.toOwnedSlice(allocator),
    };
}

fn isFresh(id: u64, ranges: []const Range) bool {
    for (ranges) |r| {
        if (r.contains(id)) {
            return true;
        }
    }
    return false;
}

fn part1(input: []const u8) !usize {
    const allocator = std.heap.page_allocator;
    const parsed = try parseInput(allocator, input);
    defer allocator.free(parsed.ranges);
    defer allocator.free(parsed.ids);

    var count: usize = 0;
    for (parsed.ids) |id| {
        if (isFresh(id, parsed.ranges)) {
            count += 1;
        }
    }

    return count;
}

fn lessThan(_: void, a: Range, b: Range) bool {
    return a.start < b.start;
}

fn mergeRanges(allocator: std.mem.Allocator, ranges: []const Range) ![]Range {
    if (ranges.len == 0) {
        return try allocator.alloc(Range, 0);
    }

    // Sort ranges by start position
    const sorted = try allocator.alloc(Range, ranges.len);
    @memcpy(sorted, ranges);
    std.mem.sort(Range, sorted, {}, lessThan);

    var merged = try std.ArrayList(Range).initCapacity(allocator, 10);
    errdefer merged.deinit(allocator);

    try merged.append(allocator, sorted[0]);

    for (sorted[1..]) |r| {
        var last = &merged.items[merged.items.len - 1];

        // Check if ranges overlap or are adjacent
        if (r.start <= last.end + 1) {
            // Merge by extending the end
            if (r.end > last.end) {
                last.end = r.end;
            }
        } else {
            // No overlap, add as new range
            try merged.append(allocator, r);
        }
    }

    allocator.free(sorted);
    return merged.toOwnedSlice(allocator);
}

fn countIDsInRanges(allocator: std.mem.Allocator, ranges: []const Range) !u64 {
    const merged = try mergeRanges(allocator, ranges);
    defer allocator.free(merged);

    var total: u64 = 0;
    for (merged) |r| {
        total += r.end - r.start + 1;
    }
    return total;
}

fn part2(input: []const u8) !u64 {
    const allocator = std.heap.page_allocator;
    const parsed = try parseInput(allocator, input);
    defer allocator.free(parsed.ranges);
    defer allocator.free(parsed.ids);

    return try countIDsInRanges(allocator, parsed.ranges);
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    const file = try std.fs.cwd().openFile("../inputs/day05.txt", .{});
    defer file.close();

    const input = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(input);

    const result1 = try part1(input);
    const result2 = try part2(input);

    std.debug.print("Part 1: {}\n", .{result1});
    std.debug.print("Part 2: {}\n", .{result2});
}

test "part1" {
    const example =
        \\3-5
        \\10-14
        \\16-20
        \\12-18
        \\
        \\1
        \\5
        \\8
        \\11
        \\17
        \\32
    ;

    const result = try part1(example);
    try std.testing.expectEqual(@as(usize, 3), result);
}

test "part2" {
    const example =
        \\3-5
        \\10-14
        \\16-20
        \\12-18
        \\
        \\1
        \\5
        \\8
        \\11
        \\17
        \\32
    ;

    const result = try part2(example);
    try std.testing.expectEqual(@as(u64, 14), result);
}

File: rust/src/bin/day05.rs
use std::fs;

#[derive(Debug, Clone, Copy)]
struct Range {
    start: u64,
    end: u64,
}

impl Range {
    fn contains(&self, id: u64) -> bool {
        id >= self.start && id <= self.end
    }
}

fn parse_input(input: &str) -> (Vec<Range>, Vec<u64>) {
    let parts: Vec<&str> = input.split("\n\n").collect();

    let ranges: Vec<Range> = parts[0]
        .lines()
        .map(|line| {
            let parts: Vec<&str> = line.split('-').collect();
            Range {
                start: parts[0].parse().unwrap(),
                end: parts[1].parse().unwrap(),
            }
        })
        .collect();

    let ids: Vec<u64> = parts[1]
        .lines()
        .map(|line| line.parse().unwrap())
        .collect();

    (ranges, ids)
}

fn is_fresh(id: u64, ranges: &[Range]) -> bool {
    ranges.iter().any(|range| range.contains(id))
}

fn part1(input: &str) -> usize {
    let (ranges, ids) = parse_input(input.trim());

    ids.iter()
        .filter(|&&id| is_fresh(id, &ranges))
        .count()
}

fn merge_ranges(ranges: &[Range]) -> Vec<Range> {
    if ranges.is_empty() {
        return Vec::new();
    }

    let mut sorted = ranges.to_vec();
    sorted.sort_by_key(|r| r.start);

    let mut merged = vec![sorted[0]];

    for range in sorted.iter().skip(1) {
        let last = merged.last_mut().unwrap();

        // Check if ranges overlap or are adjacent
        if range.start <= last.end + 1 {
            // Merge by extending the end
            last.end = last.end.max(range.end);
        } else {
            // No overlap, add as new range
            merged.push(*range);
        }
    }

    merged
}

fn count_ids_in_ranges(ranges: &[Range]) -> u64 {
    let merged = merge_ranges(ranges);
    merged.iter().map(|r| r.end - r.start + 1).sum()
}

fn part2(input: &str) -> u64 {
    let (ranges, _ids) = parse_input(input.trim());
    count_ids_in_ranges(&ranges)
}

fn main() {
    let input = fs::read_to_string("../inputs/day05.txt").expect("Failed to read input file");

    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE: &str = "\
3-5
10-14
16-20
12-18

1
5
8
11
17
32";

    #[test]
    fn test_part1() {
        assert_eq!(part1(EXAMPLE), 3);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(EXAMPLE), 14);
    }
}
